### PR TITLE
Liquidation Bot

### PR DESCRIPTION
**Pull request by Alexsey**
* Move `dotenv` and partially `yargs` setup out of the main process to be preformed before the actual program starts (it is the recommended way when working Modules i.e. `import`)
* Add TypeScript paths so folders can be imported like `import { config } from "@config"`
* Create config folder to be used across the application instead of process.env
* Move TypeScript specific environment variables from `start` script into tsconfig.json
* Add JestJS tests setup
* Extraction of Liquidation Bot script to a separate files structure
* Update dependencies to the latest versions

---
This change is [Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/8)

### CODE REVIEW

**Review by ilya-bobyr — CHANGES_REQUESTED** [review]
I really like how you structure the code.
Small files and most of the code makes sense in isolation from the rest of the system.

> Reviewed 37 of 37 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-:-MqNdTnG1Q8qt0MplFzV:b-9mexl)* status: all files reviewed, 20 unresolved discussions (waiting on @Alexsey)

___
*[.env.test, line 10 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNSKWvED8-_iDl6Reg:-MqNSKWwD1ZHdlM89OjV:bygcvp4) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/.env.test#L10)):*
> ```
> LIQUIDATION_BOT_TRADERS_FETCHER_RE_FETCH_INTERVAL_SEC=0.1
> LIQUIDATION_BOT_TRADERS_CHECKER_RECHECK_INTERVAL_SEC=0.05
> ```

_minor_

`RE_FETCH` and `RECHECH` should use the same style.
Either `re-check` and `re-fetch`, or `recheck` and `refetch`.
___
*[jest.config.js, line 9 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNSfrD6xmeLTf9yv8Q:-MqNSfrD6xmeLTf9yv8R:b-22ny6v) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/jest.config.js#L9)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
>   moduleNameMapper: {
>     "^@liquidationBot/(.*)$": "<rootDir>/src/liquidationBot/$1",
>     "^@generated/(.*)$": "<rootDir>/src/generated/$1",
>     "^@config$": "<rootDir>/src/config/config.ts",
>   },
> ```
</details>


_minor_

Would be good to add a comment clarifying that this list is also duplicated in `tsconfig.json` as `compilerOptions.paths` and we want to keep them in sync.
___
*[README.md, line 7 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNTEYu1rBD9_Y2ZLIR:-MqNTEYu1rBD9_Y2ZLIS:bonhwxx) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/README.md#L7)):*
> ```Markdown
> ### Node.js
> 
> Be sure that you have [NodeJS](https://nodejs.org/en/) and [yarn](https://yarnpkg.com/getting-started/install) installed
> ```

<style>
Sentences end with dots.

I also think that we should keep the text wrapped at some reasonable text width.
Seems like this file is wrapped on both 80 and 100 character boundaries.
Though it is inconsistent.
But this like is over both of these limits.
</style>
___
*[README.md, line 30 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNU2ZMEywjalJtVE3e:-MqNU2ZMEywjalJtVE3f:b-z20ntb) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/README.md#L30)):*
> ```Markdown
> 
> You can keep parameters for multiple chains in your `.env` file. You are going to select a specific
> chain parameters using a `--networkId` argument of the CLI commands.
> ```

We should unify our internal and external argument names.
Internally, we are using `--network` and `MAINNET_ARBITRUM`.
I'll send a PR to fix it.
___
*[src/index.ts, line 284 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNUf226aezjKrGhZGL:-MqNUf226aezjKrGhZGM:b-5horci) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/index.ts#L284)):*
> ```XML
>           ? liquidationBotReporting.pm2.start(liquidationBot)
>           : liquidationBotReporting.console.start(liquidationBot),
>         // new Promise((r) => setTimeout(r, 5_000)).then(liquidationBot.stop),
> ```

Remove or document commented out code.
___
*[src/config/config.ts, line 37 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNVHEREsvseTeOGTEZ:-MqNVHEREsvseTeOGTE_:be3tw2r) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/config/config.ts#L37)):*
> ```XML
> 
> const network =
>   getStringEnv("NETWORK_ID").toUpperCase() == "ARBITRUM_RINKEBY"
> ```

This check should not be just "`ARBITRUM_RINKEBY` or anything else".
We should validate specific values and provide an error for all other cases.
There is a `getEnumEnv` helper.
Maybe we could use it here?
___
*[src/config/config.ts, line 67 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNUstH9oYRW86-s9lD:-MqNUstH9oYRW86-s9lE:b-keu5xh) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/config/config.ts#L67)):*
> ```XML
>   exchangeAddress,
>   tradesUrl: TRADES_URL_PREFIX + exchangeAddress,
>   walletPath: WALLET_PREFIX + getStringEnv("ACCOUNT_NUMBER"),
> ```

It would be more accurate to use `getNumberEnv` for `ACCOUNT_NUMBER`.
Ideally, it should be a positive integer.

I already have a follow-up PR to make `getNumberEnv` validate values that are expected to be number and show meaningful errors.
___
*[src/config/yargsPreLoader.ts, line 35 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNVdo6E9jFIWIJBt3k:-MqNVdo6E9jFIWIJBt3l:bq8thvl) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/config/yargsPreLoader.ts#L35)):*
> ```
> const accountNumber = argv.accountNumber ?? getNumberEnv("ACCOUNT_NUMBER") ?? 0;
> if (accountNumber < 0 || accountNumber >= 200) {
> ```

If `argv.accountNumber` value is not a valid integer `(accountNumber < 0 || accountNumber >= 200)` could be `false`.
This is a better check:

```typescript
const accountNumber = Number.parseInt(argv.accountNumber) ?? getNumberEnv("ACCOUNT_NUMBER") ?? 0;
if (Number.isNaN(accountNumber) || accountNumber < 0 || accountNumber >= 200) {
```

Assuming we also merge my change, that ensures that `getNumberEnv` does number validation as well.
___
*[src/config/yargsPreLoader.ts, line 42 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNWsvQE5bFRR01o9iJ:-MqNWsvR71c3BvZlcd4A:b-idjs80) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/config/yargsPreLoader.ts#L42)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> if (argv.networkId) {
>   process.env.NETWORK_ID = argv.networkId;
> } else if (!process.env.NETWORK_ID) {
>   process.env.NETWORK_ID = "arbitrum_rinkeby";
> }
> ```
</details>


I do not think we should have a default network.
If not defined, it is probably better if it fails.
___
*[src/liquidationBot/bot.ts, line 29 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNX3o50vuE0aNQvRLC:-MqNX3o50vuE0aNQvRLD:bkxthgy) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/bot.ts#L29)):*
> ```XML
> type EventsIterator = AsyncIterableIterator<
>   [
>     | { type: "error"; error: FetchError | CheckError | LiquidationError }
> ```

I wonder if it is a good idea to use a reserved keyword as a property name.
I assume this is some kind of common pattern in TS, used to defined enumerations?

For example, the following will fail, right?

```typescript
    const events: EventsIterator = f();
    for (const event of events) {
        const { type } = event;
    }
```

`const { type }` would be a syntax error, no?
It did not work for me with the `default` keyword.
___
*[src/liquidationBot/__tests__/integration/liquidationBot.test.ts, line 7 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNY2Xp0qgTHGHTljFQ:-MqNY2Xp0qgTHGHTljFR:bywencx) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/__tests__/integration/liquidationBot.test.ts#L7)):*
> ```XML
> import { liquidationBot } from "@liquidationBot/bot";
> 
> jest.disableAutomock();
> ```

Would be great to add a comment explaining why do we do this.
___
*[src/liquidationBot/__tests__/integration/liquidationBot.test.ts, line 31 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNYIxzCbUvaQH0v5le:-MqNYIxzCbUvaQH0v5lf:b-15kpkf) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/__tests__/integration/liquidationBot.test.ts#L31)):*
> ```XML
> 
> describe("liquidationBot", () => {
>   /* Because bot processors cycles are running on timers, the most natural way
> ```

_minor_

I think the most widespread comment style for multiline comments is like this:

```typescript
  /*
   * First line contains only the start comment sequence with no text.
   * At least for the multiline comments.
   * And the very last line also contains just the end comment sequence.
   */
```

There are other styles, but, I think, we mostly use this one.
___
*[src/liquidationBot/__tests__/integration/liquidationBot.test.ts, line 32 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNZMPe5_jmNerjWKKr:-MqNZMPe5_jmNerjWKKs:blc1hv9) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/__tests__/integration/liquidationBot.test.ts#L32)):*
<details><summary><i alt="summary">Quoted 11 lines of code…</i></summary>

> ```
>    * to write integration tests on it would be to use Jest's fake timers API.
>    * But unfortunately it doesn't work properly with Promises.
>    * //https://github.com/facebook/jest/issues/7151
>    * Workaround exists, and I have tried some of them, but the results was from
>    * not working at all to working unreliable.
>    * Maybe it can be setup better, but I have decided just to set cycles time
>    * length to a very low values in .env.test and control bot execution by
>    * listening events. And it works great
>    * */
>   type EventTypes = LiquidationBotEvents["type"];
>   let botEvents: LiquidationBotEvents[] = [];
> ```
</details>


It is really great that you managed to find a way to test this!
___
*[src/liquidationBot/processors/tradersChecker.ts, line 29 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNZwKV9LaF9QUnNp8C:-MqNZwKV9LaF9QUnNp8D:b60qqrd) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/processors/tradersChecker.ts#L29)):*
> ```XML
> 
>     if (traders.length) {
>       tradersEvents.emit("gotSome", true);
> ```

_minor_

`gotSome` sounds a bit too informal :)
Maybe just `ready`?
___
*[src/liquidationBot/processors/tradersFetcher.ts, line 3 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqN_HDjCThPduSfD08G:-MqN_HDjCThPduSfD08H:b-2da8cu) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/processors/tradersFetcher.ts#L3)):*
> ```XML
> import { setTimeout } from "node:timers/promises";
> import { Readable } from "node:stream";
> import { config as appConfig } from "@config";
> ```

Why rename `config` as `appConfig`?
It does not seem to be conflicting with anything.
Renames have readability and maintenance cost.
___
*[src/liquidationBot/processors/tradersLiquidator.ts, line 55 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqN_hRoC3_e6c1jztfA:-MqN_hRoC3_e6c1jztfB:b-p6cw9e) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/processors/tradersLiquidator.ts#L55)):*
<details><summary><i alt="summary">Quoted 15 lines of code…</i></summary>

> ```
>       yield { liquidationsResults, liquidationsErrors };
> 
>       if (liquidationsErrors.length) {
>         await setTimeout(processorConfig.retryIntervalSec);
>       }
> 
>       const erroredTraders = liquidationsErrors.map(({ trader }) => trader);
>       const { nonLiquidatableTraders, liquidatableChecksErrors } =
>         await filterNonLiquidatableTraders(erroredTraders);
>       nonLiquidatableTraders.forEach((nonLiquidatableTrader) => {
>         liquidatableTraders.delete(nonLiquidatableTrader);
>       });
>       if (liquidatableChecksErrors.length) {
>         yield { liquidatableChecksErrors };
>       }
> ```
</details>


I remember we discussed the need for this block for quite some time when we did the initial review.
It would be good to have a comment here, considering it turned out to be non-trivial.
Does not have to be extremely long.
IIRC, one point was that the liquidator needs to recheck in case of an error because the priority and the timeout for this check is different from the priority and timeouts for the normal trader checking flow.
___
*[src/liquidationBot/processors/tradersLiquidator.ts, line 89 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNaJAj5eXLCXjPF5s2:-MqNaJAj5eXLCXjPF5s3:b-j5ff9d) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/processors/tradersLiquidator.ts#L89)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>     return {
>       nonLiquidatableTraders,
>       liquidatableChecksErrors,
>     };
> ```
</details>


General comment.

I wonder if some of the names could be made shorter without the loss of readability.
Maybe even increasing the readability.
In particular, names, generally, are considered in the context where they are used.

When I call a function called `filterNonLiquidatableTraders()`, I would read properties of the object it returns relative to the function name.
So, something like `traders` and `errors` could be as informative as `nonLiquidatableTraders` and `liquidatableChecksErrors`.
And if they are not clear or are ambitious in the caller context, they can be renamed there.

Sometimes it can cause problems, if the names are too generic.
And it also depends on the languages support.
I do not have as much experience with JS or TS.
So, here a different set of empirical rules should be followed, compared to Rust, Haskell or C++.
___
*[src/liquidationBot/reporting/reporterRunner.ts, line 9 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNbTmF8EPiYVFINr9Q:-MqNbTmF8EPiYVFINr9R:bun9la6) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/reporting/reporterRunner.ts#L9)):*
> ```
>     // potentially reporters may also export async init() function
>     //   Run it here and in catch below, above the reportEvent call
> ```

<style><br>
First letter in a sentence should be upper case.  And sentences should end with dots.<br>
Also, unnecessary whitespace at the beginning of the second sentence.<br>
</style>
___
*[src/liquidationBot/reporting/reporterRunner.ts, line 14 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNbscr-08jtCzSpTN_:-MqNbscs7y3J6zhYKYh4:b-ht9q9e) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/reporting/reporterRunner.ts#L14)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
>     // @ts-ignore - cause is a new feature that would be typed in TS 4.5 https://github.com/microsoft/TypeScript/pull/46291
>     const reporterCrashError = Error("Reporter will be restarted after crash", {
>       cause: error,
>     });
> ```
</details>


`cause` is a really nice feature to have.
___
*[src/liquidationBot/reporting/reporters/pm2.ts, line 101 at r1](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MqNczZh2x7zHAD5SovE:-MqNczZiETBrpBrmUoCe:b-rx0vlj) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/7e35a1ece1d866252e1140075302275e6f0772bc/src/liquidationBot/reporting/reporters/pm2.ts#L101)):*
> ```
> // e.g. 30s, 1m 30s, 1h, 1h 59m, 17d 15h
> function getDurationStr(from: number | undefined, to = Date.now()) {
> ```

Pretty cool implementation.

Alternatively, `date-fns`, which seems to be a very widely used library, provides `formatDuration`: https://date-fns.org/v2.27.0/docs/formatDuration

**Comment by Alexsey** [comment]
@ilya-bobyr 

Regarding usage of `type` - it is not a reserved word so it's absolutely safe to use it as a field name. `default` is also safe to use a field name but is an early error to use as a variable name. This is why you was not able to do `const { default } = ...` but `const { default: def } = ...` would had work

I cannot tell that using specifically "type" the purpose of type distinguishing is a pattern - it would probably be ok to use a synonym, but "type" is definitely used for this purpose quite often

===

Regarding `jest.disableAutomock();` - the reason for disabling automock is because its an integration test. So adding a comment here would require do duplicate it to every future integration test

All integration tests would probably prefer to disable automock because their goal is to test up to the service boundaries. So using automock would require to unmock almost all internal structures of the service

===

Regarding making manes shorter in the `liquidationsGenerator` function - I would be glad to use names like `traders` and `errors` but in this specific context I already have `liquidatableTraders`, `liquidatedTraders` and `erroredTraders` for `traders` and `liquidationsErrors` with `liquidatableChecksErrors` for `errors` so having just `traders` or `errors` may have sense as the return value of some function but become ambiguity one line above, when it's getting used in the mix with other types of it kind

There is nothing specific to TS or JS - I just think that names like "id" should be used only until there is only one type of ids in the context. But when another "id" appeares they should have distinguishable names for not having "id" and "userId" at the same time

===

Regarding usage of `date-fns` - probably I've just missed it :) But unfortunately in this place I'm following the format PM2 is already using and `date-fns` doesn't allow to format date this way

Of course, everywhere we can we must use tools that we decide to have in our toolset

**Comment by Alexsey** [comment]
The rest of the comments has been resolved

File: src/liquidationBot/services/traders.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import axios from "axios";
+import { config } from "@config";
+import type { Trader } from "@liquidationBot/types";
+
+const { tradesUrl } = config;
+
+export const getOpen = async (): Promise<Trader[]> => {
+  const { data } = await axios.get<{ trades: { trader: Trader }[] }>(tradesUrl);
+  return data.trades.map(({ trader }) => trader);
```
Comment (kienswap) [inline src/liquidationBot/services/traders.ts:9]: Shouldn't this deduplicate the traders list? This seems to contain a lot of duplicates if it maps trade -> trader 1:1.

**Review by ilya-bobyr — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, 1 of 2 files at r4, 3 of 3 files at r5, 3 of 3 files at r9, 3 of 3 files at r10.
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/8)* status: 35 of 37 files reviewed, 10 unresolved discussions (waiting on @Alexsey and @ilya-bobyr)

**Review by ilya-bobyr — COMMENTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/8)* status: all files reviewed, 10 unresolved discussions (waiting on @Alexsey)

**Review by ilya-bobyr — COMMENTED** [review]
> Dismissed @kienswap from a discussion.
> *[Reviewable](https://reviewable.io/reviews/futureswap/fs-cli-v4/8)* status: all files reviewed, 9 unresolved discussions (waiting on @Alexsey)

___
*[src/liquidationBot/services/traders.ts, line 9 at r10](https://reviewable.io/reviews/futureswap/fs-cli-v4/8#-MnqycVn363VUfBq2NkK-r10-9:-MqrqNkQDQeFERGpU7-Y:b-fxsyb) ([raw file](https://github.com/futureswap/fs-cli-v4/blob/0190cd11bce92ec2f83482ef06c2e78ff6c4662c/src/liquidationBot/services/traders.ts#L9)):*
<details><summary><i>Previously, kienswap wrote…</i></summary><blockquote>

Shouldn't this deduplicate the traders list? This seems to contain a lot of duplicates if it maps trade -> trader 1:1.
</blockquote></details>

Trader indexer returns a list of open trades.
So there are no duplicates.

It would be a bad idea to return a list of all trades :)
It would grow over time.  And we do not really need it here in the first place.

**Comment by ilya-bobyr** [comment]
Merged this change in in https://github.com/futureswap/fs-cli-v4/pull/11

**Pull request closed without merging**